Added the two new types any_pointer and any_array

Tisten opened the pull request:
As requested in the issue https://github.com/wc-duck/datalibrary/issues/145

Commit:
Added the two new types anyptr and anyarray

As requested in the issue https://github.com/wc-duck/datalibrary/issues/145

Tisten commented on src/dl.cpp:
```diff
@@ -43,9 +43,9 @@ dl_error_t dl_context_destroy(dl_ctx_t dl_ctx)
-                             void*                instance,        size_t instance_size,
-                             const unsigned char* packed_instance, size_t packed_instance_size,
-                             size_t*              consumed )
+							 void*                instance,        size_t instance_size,
```
This PR contains some "space to tab" conversions to minimize the "strange indent here" feedback.

wc-duck replied:
I'll try to get to fixing a clang-format file :)

Tisten commented on src/dl.cpp:
```diff
@@ -243,18 +243,78 @@ static dl_error_t dl_internal_store_array( dl_ctx_t dl_ctx, dl_type_storage_t st
+			return err;
+		dl_binary_writer_seek_set( &store_ctx->writer, sub_pos );
+								
+		// use when fast_loading is merged 
```
This code was written to work with fast pointer patching, but since that PR hasn't been merged I've re-written this PR to work without it.

Tisten commented on src/dl_typelib_write_c_header.cpp:
```diff
@@ -201,6 +223,35 @@ static void dl_context_write_c_header_begin( dl_binary_writer* writer, const cha
+									   "   // ... DL_DECLARE_ANY ...\n"
+									   "	   typedef uint32_t dl_typeid_t;\n"
+									   "\n"
+									   "	   struct dl_anyptr {\n"
```
This is the C/C++ type for anyptr

Tisten commented on src/dl_typelib_write_c_header.cpp:
```diff
@@ -481,7 +532,7 @@ static void dl_context_write_c_header_enums( dl_binary_writer* writer, dl_ctx_t
 }
 
-static void dl_context_write_type( dl_ctx_t ctx, dl_type_storage_t storage, dl_typeid_t tid, dl_binary_writer* writer )
+static dl_error_t dl_context_write_type( dl_ctx_t ctx, dl_type_storage_t storage, dl_typeid_t tid, dl_binary_writer* writer )
```
This PR also dependended on [the ](https://github.com/wc-duck/datalibrary/pull/159), and it was to much work to rewrite all code to "not" do proper error handling. So now the error handling is so-so in the PR, and many lines like these are a duplication of that PR. Please try to ignore everything which deals with error propagation and indentation because of the changed return type in the function signature

Commit:
Cleaned up the review a bit

Commit:
Merge remote-tracking branch 'origin/master' into anyptr_and_anyarray

wc-duck commented on README.md:
```diff
@@ -60,16 +60,18 @@ tool\
+| inline_array   | int8[20], some_type[20]      | type[20]                                    | fixed length, type can be any POD-type or user defined struct        |
+| array          | int8[], some_type[]          | struct { type* data; uint32_t count }       | dynamic length, type can be any POD-type or user defined struct      |
+| pointer        | some_type*                   | type*                                       | pointer to any user-defined type                                     |
+| anyptr         | anyptr                       | struct { void* ptr; dl_typeid_t tid; }      | pointer to any struct type, where the type is decided on assignment  |
```
for consistency this should probably be 'any_pointer'... and I wished I would have called 'pointer' 'ptr', but the ship has sailed on that one!
same with 'any_array' is consistent with 'inline_array'

wc-duck replied:
and should there be an 'any_inline_array' for completeness?

Tisten replied:
Renamed them
any_inline_array needs a known element (max)size, like a union. It would be better to just change the union type to allow explicit max size and assignment of any type to it in that case

wc-duck commented on README.md:
```diff
@@ -273,6 +279,22 @@ left untouched!",
         // inline arrays are just arrays with the correct type ...
         "inline_array_member1"  : [1, 2, 3],
+		
+		// an anyptr can be null, and then it is typeless
```
I'll complain about indent here as this will not be fixable by a clang-format :)

wc-duck commented on README.md:
```diff
@@ -273,6 +279,22 @@ left untouched!",
+		// an anyptr can be null, and then it is typeless
+		"anyptr_member1": null,
+		// but when it has a value then that will be stored in the "__subdata" section, and there will be a "__any_type_" entry right after the member.
+		"anyptr_member2": "ptr1234"
```
This format will not really work if you have an array[any_pointer] or inline_array[any_pointer]... I would probably go with:
```json
"anyptr_member2" : {
    "type" : "ptr1234",
    "data" : {
        "x" : 1234,
        "y" : 5678
    }
}
```

that would also mesh better with any_array as that would be:
```json
"anyarray_member" : {
    "type" : "the_type",
    "data" : [ { "x" : 1, "y" : 2 }, { "x" : 4, "y" : 5 } ]
}
```

Tisten replied:
Well arrays and inline arrays are functionally working today as you see in the tests, but I agree that they are not very pretty, and you can't put the elements wherever since the parser require them to be in order which might not be what you expect. So I might look into that when I get back to improving data_library. I've started on other tasks right now, so I will leaved as is this for a time now.

wc-duck replied:
If I find the time and energy for it I might have a go at it myself... do you agree with the proposed format above?

wc-duck replied:
We could maybe embed the type in the member-name... but that might be a bit "icky"... something like "my_any_member[the_type]" : { ... } ?

wc-duck replied:
... and it will not work for array-of-anyptr... back to the initial suggestion.

Tisten replied:
I do like your initial suggestion.

Tisten replied:
Fixed now

wc-duck commented on include/dl/dl.h:
```diff
@@ -136,6 +136,8 @@ typedef enum
 	DL_TYPE_STORAGE_STR,
 	DL_TYPE_STORAGE_PTR,
 	DL_TYPE_STORAGE_STRUCT,
+	DL_TYPE_STORAGE_ANYPTR,
```
shouldn't the storage here be just DL_TYPE_STORAGE_ANY and that combined with DL_TYPE_ATOM above would give you if it is an any_pointer, any_array or any_inline_array.

Tisten replied:
Using storage prevented me from having anyarray[] which I really wanted without an extra struct in there.
The "any_inline_array" can't happen, since the "any" is of arbitrary size, so it needs a pointer (arrays use pointers) and thus can't be stored inline. If you want to store "any" inline then you must have a "size" and you can only store types smaller or equal to that size. Which basically makes it a union which you already support.

wc-duck replied:
You are ofcourse correct... I thought an inline array of anyptr... but that should just be "anyptr[16]" i guess

Tisten replied:
Indeed, as shown with the type `WithInlineAnyptrArray` in the tests

wc-duck commented on src/dl.cpp:
```diff
@@ -247,18 +247,78 @@ static dl_error_t dl_internal_store_array( dl_ctx_t dl_ctx, dl_type_storage_t st
 }
 
-static dl_error_t dl_internal_store_member( dl_ctx_t dl_ctx, const dl_member_desc* member, uint8_t* instance, CDLBinStoreContext* store_ctx )
+static dl_error_t dl_internal_store_anyarray( dl_ctx_t dl_ctx, const uint8_t* data_ptr, uint32_t count, CDLBinStoreContext* store_ctx )
```
I'm not sure, but if you split the DL_STORAGE_TYPE mentioned above into just using _ANY and use the atom type as well I think this might turn out a bit cleaner?

Tisten replied:
Explained here https://github.com/wc-duck/datalibrary/pull/164#discussion_r1012206239

wc-duck commented on src/dl.cpp:
```diff
@@ -247,18 +247,78 @@ static dl_error_t dl_internal_store_array( dl_ctx_t dl_ctx, dl_type_storage_t st
 	dl_type_storage_t storage_type = member->StorageType();
 
+	if (storage_type == DL_TYPE_STORAGE_ANYARRAY)
+		atom_type = DL_TYPE_ATOM_ARRAY; // To avoid code duplification and simplify implementation
```
:point_up:

Tisten replied:
Would you prefer the code duplification here?

wc-duck reviewed (commented):
I'm going to wait with more input until fast-patching is merged and the DL_STORAGE_TYPE_ANY is at least investigated or explained to me why it is a bad idea if it is :)

Commit:
Merge remote-tracking branch 'origin/master' into anyptr_and_anyarray

Commit:
Renamed anyptr to any_pointer and anyarray to any_array

Commit:
Fixed the docs as well

Commit:
Whitespace in README

Commit:
Merge branch 'master' into anyptr_and_anyarray

Commit:
Merge branch 'official_master' into anyptr_and_anyarray

Commit:
Merge branch 'master' into anyptr_and_anyarray

Commit:
Redid the any_pointer and any_array json format

The new style is
"MyAnyPointer" : {
    "type": "the_type",
    "data": "ptr_88"
},
"MyAnyArray" : {
    "type": "the_type",
    "data": [ { "member": "value" } ]
}

Commit:
And the implementation for the last change

Commit:
Gcc giving other warnings than msvc

Commit:
And one more warning

Commit:
Do I need to get WSL?

Tisten committed:
Merge branch 'Official_master' into anyptr_and_anyarray

# Conflicts:
#	src/dl.cpp
#	src/dl_txt_pack.cpp
#	src/dl_typelib_write_c_header.cpp
#	tests/dl_tests_error.cpp

Tisten committed:
Fixed bad merge

Tisten committed:
Different warnings from cmake and bam
